### PR TITLE
enhancement: mock_db_session fixture

### LINKED ISSUE #94: enhancement: async postgres(-like) DB fixtures for unit tests
Working on #93, I've spent enough time banging my head into pytest walls that I'm pulling it out into its own issue. 

# Goals: 
  1. the DB fixture should be lightweight / not impose a significant performance penalty. 
    - Less concerned with local development than I am with slowing down the CI pipeline. We've got a pretty ergonomic setup for local testing against a proper postgres instance already with docker compose etc, I just want unit tests for coverage's sake.  
  2. The fixture should "just work" with our async code / tests.
  3. The fixture should "just work" with the postgres SQLAlchemy dialect  
  4. The fixture should target the fastAPI dependency injection system -- where possible I think we want to avoid mocking things at the level of individual routes. 

The current approach uses an in-memory SQLite DB by setting `mock_settings.SQLALCHEMY_DATABASE_URL = "sqlite+aiosqlite:///:memory:"`. 
### the Good about the current approach:
  - sqlite+aiosqlite is lightweight and async
  - The fixtures I've hacked together so far work by overriding the `get_settings` dependency and the `get_db_session` dependency, which helps keep the tests exercising the routes pretty simple. 
  
tl;dr we've currently got 1, ~2, and 4 from above 
### the Bad: 
  - doesn't play nicely when we use postgres-specific features (e.g. the `ARRAY(String)` type we're using for tags) 
  - A reasonable attempt to hack the postgres-specific array type ([gist](https://gist.github.com/OwenPriceSkelly/a7c30cb9d1cfded6ae9686932e03e3f5)) causes weird async/io errors. I ran into similar errors in the actual implementation, but those errors were fixed by setting the cascade settings correctly for entrypoint relationships. 
    - I'm not sure if these new errors means there's some kind of problem with the fake array type directly or if there's some inconsistency with the async tests / fixture database (of course also possible that I didn't *really* fix the real code, but it works fine with real data against a local postgres instance so I don't think that's the case).

### PR DESCRIPTION
**Pull request by hholb**
Resolves: #94

## Overview

This PR sets up the test suite to run integration tests alongside a postgres instance provided by the [testcontainers](https://testcontainers-python.readthedocs.io/en/latest/) library. Tests that need to interact with the database can use the `mock_db_session` fixture defined in `conftest.py`. The `mock_db_session` fixture handles overriding the `get_db_session` dependency used by api endpoints and provides each test with a clean database to run queries on.

This PR also restructures the test sure to mirror the structure of the `src/` directory.
```text
.
└── tests
    ├── api
    │   └── routes
    │       ├── test_docker_push_token.py
    │       ├── test_doi.py
    │       ├── test_entrypoint.py
    │       ├── test_garden.py
    │       ├── test_greet.py
    │       ├── test_notebook.py
    │       └── test_routes.py
    ├── auth
    │   └── test_globus_groups.py
    ├── conftest.py
    ├── fixtures
    │   ├── EntrypointCreateRequest-shared-entrypoint.json
    │   ├── EntrypointCreateRequest-with-metadata.json
    │   ├── GardenCreateRequest-shares-entrypoint.json
    │   └── GardenCreateRequest-two-entrypoints.json
    ├── __init__.py
    └── README.md
```

## Discussion

The `mock_db_session` fixture depends on docker being available in the environment the tests are running in. To disable these tests run `pytest -m "not integration"`.  Any new tests that use the `mock_db_session` fixture should add a `@pytest.mark.integration` decorator. If docker is not present in the environment the tests that use `mock_db_session` are automatically skipped.

I had issues running the GitHub Actions workflow while testing. Steps running after the `abatilo/actions-poetry@v3` action could not find poetry on the PATH so I changed the workflow to install poetry via pipx and add poetry to PATH manually.

## Testing
Run the test suite and behold:
```bash
pytest

pytest -m "integration"

pytest -m "not integration"
```

Run the ci workflow locally with [act](https://github.com/nektos/act): `act -j build`

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -21,14 +22,26 @@ jobs:
         with:
           python-version: "3.10"
 
-      - name: Install poetry
-        uses: abatilo/actions-poetry@v3
+      - name: Install pipx
```
Comment (OwenPriceSkelly) [inline .github/workflows/ci.yaml:25]: good catch on this! The `abatilo/actions-poetry@v3` claims to already install with pipx, wonder why it was breaking. Maybe something to do with `$GITHUB_PATH` variable? could be worth opening an issue for the maintainer if there isn't one already

File: garden-backend-service/.flake8 [hunk 2]
```diff
@@ -1,3 +1,4 @@
 [flake8]
 max-line-length = 150
-extend-ignore = E203
+extend-ignore = E203, PT004
+exclude = migrations,__pycache__,schemas
```
Comment (OwenPriceSkelly) [inline garden-backend-service/.flake8]: why ignore schemas? (assuming this matches `src/api/schemas`)

File: garden-backend-service/src/api/routes/entrypoint.py [hunk 3]
```diff
@@ -22,6 +22,7 @@ async def add_entrypoint(
     db.add(new_entrypoint)
     try:
         await db.commit()
+        await db.refresh(new_entrypoint)
```
Comment (OwenPriceSkelly) [inline garden-backend-service/src/api/routes/entrypoint.py]: this seems reasonable - was this necessary to fix a bug with the tests or just better practice?

File: garden-backend-service/tests/README.md
Comment (OwenPriceSkelly) [inline garden-backend-service/tests/README.md:1]: love the readme!!

File: garden-backend-service/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    _get_auth_token,
+)
+
+from src.config import Settings, get_settings
+from src.main import app
+from src.models.base import Base
+
+
+@pytest.fixture
+def client(patch_globus_groups):
```
Comment (OwenPriceSkelly) [inline garden-backend-service/tests/conftest.py:25]: understanding check: what's the difference between this client and the previous `from fastapi.testclient import TestClient`?

File: garden-backend-service/tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    mocker.patch("src.api.dependencies.auth.add_user_to_group")
+
+
+def docker_available():
+    available = shutil.which("docker")
+    return available
+
+
+def pytest_collection_modifyitems(session, config, items):
+    """Skip integration tests that rely on docker if docker is not availablie"""
```
Comment (OwenPriceSkelly) [inline garden-backend-service/tests/conftest.py]:
```suggestion
    """Skip integration tests that rely on docker if docker is not available"""
```

File: garden-backend-service/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+def patch_globus_groups(mocker):
+    mocker.patch("src.api.dependencies.auth.add_user_to_group")
+
+
+def docker_available():
+    available = shutil.which("docker")
+    return available
+
+
+def pytest_collection_modifyitems(session, config, items):
```
Comment (OwenPriceSkelly) [inline garden-backend-service/tests/conftest.py:40]: This is very slick, big fan

**Review by OwenPriceSkelly — APPROVED** [review]
LGTM! This has me excited to write tests, which is a _very_ new feeling for me. Couple of questions just for my own understanding but nothing blocking - merge away 🫡

File: garden-backend-service/.flake8 (hunk 2 above)
Comment (hholb) [reply garden-backend-service/.flake8]: Oh, whoops! I added this to my local config while working on this to suppress some flake8 complaints while I worked on the tests. I meant to take it out before pushing.

**Review by MaxTuecke — APPROVED** [review]
Very thorough tests, this looked like a fun one to figure out. As far as i can tell, it LGTM!

File: garden-backend-service/src/api/routes/entrypoint.py (hunk 3 above)
Comment (hholb) [reply garden-backend-service/src/api/routes/entrypoint.py]: I thought some of the entrypoint tests were failing due to the relationships not being updated when adding a new one to the database. Turns out that wasn't the issue and the call to refresh is unnecessary. The tests pass without it.

File: garden-backend-service/tests/conftest.py (hunk 4 above)
Comment (hholb) [reply garden-backend-service/tests/conftest.py:25]: The `fastiapi.TestClient` is built on the `httpx.Client` which only supports synchronous requests. The fastapi [docs](https://fastapi.tiangolo.com/advanced/async-tests/) recommend using the `httpx.AsyncClient` in async tests.

File: garden-backend-service/tests/conftest.py (hunk 5 above)
Comment (hholb) [reply garden-backend-service/tests/conftest.py]: Good catch!
